tough/src/editor.rs

```diff
@@ -0,0 +1,660 @@
+            field: "timestamp version",
+        })?;
+        let expires = self.timestamp_expires.context(error::Missing {
+            field: "timestamp expiration",
+        })?;
+        let _extra = self.timestamp_extra.clone().unwrap_or_else(HashMap::new);
+
+        let snapshot_meta = Self::timestamp_meta(signed_snapshot);
+
+        let mut timestamp = Timestamp::new(SPEC_VERSION.to_string(), version, expires);
+        timestamp
+            .meta
+            .insert("snapshot.json".to_owned(), snapshot_meta);
+        timestamp._extra = _extra;
+
+        Ok(timestamp)
+    }
+
+    // Build a TimestampMeta struct from a given SignedRole<T>. This metadata
+    // includes the sha256 and length of the signed role.
+    fn timestamp_meta<T>(role: &SignedRole<T>) -> TimestampMeta
+    where
+        T: Role,
+    {
+        TimestampMeta {
+            hashes: Hashes {
+                sha256: role.sha256.to_vec().into(),
+                _extra: HashMap::new(),
+            },
+            length: role.length,
+            version: role.signed.signed.version(),
+            _extra: HashMap::new(),
+        }
+    }
+
+    // Signs a given role. This function finds the appropriate keys to sign
+    // given role, signs the role, and builds a `SignedRole<T>`.
+    fn sign_role<T>(
+        &self,
+        role: T,
+        keys: &RootKeys,
+        rng: &dyn SecureRandom,
+    ) -> Result<SignedRole<T>>
+    where
+        T: Role + Serialize,
+    {
+        let root = &self.signed_root.signed.signed;
+        let role_keys = root.roles.get(&T::TYPE).context(error::NoRoleKeysinRoot {
+            role: &T::TYPE.to_string(),
+        })?;
+        let (signing_key_id, signing_key) = keys
+            .iter()
+            .find(|(keyid, _signing_key)| role_keys.keyids.contains(&keyid))
+            .context(error::SigningKeysNotFound {
+                role: &T::TYPE.to_string(),
+            })?;
+
+        let mut role = Signed {
+            signed: role,
+            signatures: Vec::new(),
+        };
+
+        let mut data = Vec::new();
+        let mut ser = serde_json::Serializer::with_formatter(&mut data, CanonicalFormatter::new());
+        role.signed
+            .serialize(&mut ser)
+            .context(error::SerializeRole)?;
+        let sig = signing_key.sign(&data, rng)?;
+        role.signatures.push(Signature {
+            keyid: signing_key_id.clone(),
+            sig: sig.into(),
+        });
+
+        let mut buffer = serde_json::to_vec_pretty(&role).context(error::SerializeSignedRole)?;
+        buffer.push(b'\n');
+        let length = buffer.len() as u64;
+
+        let mut sha256 = [0; SHA256_OUTPUT_LEN];
+        sha256.copy_from_slice(digest(&SHA256, &buffer).as_ref());
+
+        let signed_role = SignedRole {
+            signed: role,
+            buffer,
+            sha256,
+            length,
+        };
+
+        Ok(signed_role)
+    }
+}
+
+#[cfg(test)]
+mod tests {
```

Comment:
I guess I'm drawing a line, albeit a squiggly one, between the unit tests and tests that include other parts of the library.

I was thinking that the tests that create/write/load would go here since they use more than just the `Editor` library. Tests that specifically test the `Editor` library would go in the lib.